<h1>Activity 1 - PALMER PENGUINS DATA ANALYTICS PROJECT </h1>
<hr>
<h3>Analyst : Andrei Miguel P. Moralina </h3>

In [1]:
#import libraries
import numpy as np
import pandas as pd

<h3>Converting Dataset(s) into DataFrame</h3>
<h2>Requirements</h2>
<ul>
    <li>Name of the DataFrame</li>
    <li>File Path</li>
</ul>

In [2]:
penguins = pd.read_csv('datasets\\penguins.csv')

In [3]:
penguins

rowid    species     island  bill_length_mm  bill_depth_mm  \
0        1     Adelie  Torgersen            39.1           18.7   
1        2     Adelie  Torgersen            39.5           17.4   
2        3     Adelie  Torgersen            40.3           18.0   
3        4     Adelie  Torgersen             NaN            NaN   
4        5     Adelie  Torgersen            36.7           19.3   
..     ...        ...        ...             ...            ...   
339    340  Chinstrap      Dream            55.8           19.8   
340    341  Chinstrap      Dream            43.5           18.1   
341    342  Chinstrap      Dream            49.6           18.2   
342    343  Chinstrap      Dream            50.8           19.0   
343    344  Chinstrap      Dream            50.2           18.7   

     flipper_length_mm  body_mass_g     sex  year  
0                181.0       3750.0    male  2007  
1                186.0       3800.0  female  2007  
2                195.0       3250.0  female  2007  
3                  NaN          NaN     NaN  2007  
4                193.0       3450.0  female  2007  
..                 ...          ...     ...   ...  
339              207.0       4000.0    male  2009  
340              202.0       3400.0  female  2009  
341              193.0       3775.0    male  2009  
342              210.0       4100.0    male  2009  
343              198.0       3775.0  female  2009  

[344 rows x 9 columns]

<h2>Descriptive Analytics</h2>

In [4]:
penguins.describe()

rowid  bill_length_mm  bill_depth_mm  flipper_length_mm  \
count  344.000000      342.000000     342.000000         342.000000   
mean   172.500000       43.921930      17.151170         200.915205   
std     99.448479        5.459584       1.974793          14.061714   
min      1.000000       32.100000      13.100000         172.000000   
25%     86.750000       39.225000      15.600000         190.000000   
50%    172.500000       44.450000      17.300000         197.000000   
75%    258.250000       48.500000      18.700000         213.000000   
max    344.000000       59.600000      21.500000         231.000000   

       body_mass_g         year  
count   342.000000   344.000000  
mean   4201.754386  2008.029070  
std     801.954536     0.818356  
min    2700.000000  2007.000000  
25%    3550.000000  2007.000000  
50%    4050.000000  2008.000000  
75%    4750.000000  2009.000000  
max    6300.000000  2009.000000

In [5]:
penguins.describe().T

count         mean         std     min       25%      50%  \
rowid              344.0   172.500000   99.448479     1.0    86.750   172.50   
bill_length_mm     342.0    43.921930    5.459584    32.1    39.225    44.45   
bill_depth_mm      342.0    17.151170    1.974793    13.1    15.600    17.30   
flipper_length_mm  342.0   200.915205   14.061714   172.0   190.000   197.00   
body_mass_g        342.0  4201.754386  801.954536  2700.0  3550.000  4050.00   
year               344.0  2008.029070    0.818356  2007.0  2007.000  2008.00   

                       75%     max  
rowid               258.25   344.0  
bill_length_mm       48.50    59.6  
bill_depth_mm        18.70    21.5  
flipper_length_mm   213.00   231.0  
body_mass_g        4750.00  6300.0  
year               2009.00  2009.0

<h2>Checking Null Values</h2>

In [6]:
penguins.isnull()

rowid  species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0    False    False   False           False          False              False   
1    False    False   False           False          False              False   
2    False    False   False           False          False              False   
3    False    False   False            True           True               True   
4    False    False   False           False          False              False   
..     ...      ...     ...             ...            ...                ...   
339  False    False   False           False          False              False   
340  False    False   False           False          False              False   
341  False    False   False           False          False              False   
342  False    False   False           False          False              False   
343  False    False   False           False          False              False   

     body_mass_g    sex   year  
0          False  False  False  
1          False  False  False  
2          False  False  False  
3           True   True  False  
4          False  False  False  
..           ...    ...    ...  
339        False  False  False  
340        False  False  False  
341        False  False  False  
342        False  False  False  
343        False  False  False  

[344 rows x 9 columns]

In [7]:
penguins.isnull().sum()

rowid                 0
species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

<h2>Filtering a column</h2>

In [8]:
penguins['species']

0         Adelie
1         Adelie
2         Adelie
3         Adelie
4         Adelie
         ...    
339    Chinstrap
340    Chinstrap
341    Chinstrap
342    Chinstrap
343    Chinstrap
Name: species, Length: 344, dtype: object

In [9]:
penguins['species'].value_counts()

species
Adelie       152
Gentoo       124
Chinstrap     68
Name: count, dtype: int64

In [10]:
penguins['species'] == 'Chinstrap'

0      False
1      False
2      False
3      False
4      False
       ...  
339     True
340     True
341     True
342     True
343     True
Name: species, Length: 344, dtype: bool

<h2>FILLING UP NULL VALUES</h2>

In [11]:
penguins.isnull().sum()

rowid                 0
species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

In [12]:
penguins['bill_length_mm'].mean()

np.float64(43.9219298245614)

In [13]:
penguins['bill_length_mm'] = penguins['bill_length_mm'].fillna(penguins['bill_length_mm'].mean())

In [14]:
penguins.isnull().sum()

rowid                 0
species               0
island                0
bill_length_mm        0
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

In [15]:
penguins['bill_depth_mm'] = penguins['bill_depth_mm'].fillna(penguins['bill_depth_mm'].mean())
penguins['flipper_length_mm'] = penguins['flipper_length_mm'].fillna(penguins['flipper_length_mm'].mean())
penguins['body_mass_g'] = penguins['body_mass_g'].fillna(penguins['body_mass_g'].mean())

In [16]:
penguins.isnull().sum()

rowid                 0
species               0
island                0
bill_length_mm        0
bill_depth_mm         0
flipper_length_mm     0
body_mass_g           0
sex                  11
year                  0
dtype: int64

In [17]:
penguins['sex'] = penguins['sex'].fillna(penguins['sex'].mode()[0])

In [18]:
penguins.isnull().sum()

rowid                0
species              0
island               0
bill_length_mm       0
bill_depth_mm        0
flipper_length_mm    0
body_mass_g          0
sex                  0
year                 0
dtype: int64

In [19]:
penguins.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   rowid              344 non-null    int64  
 1   species            344 non-null    object 
 2   island             344 non-null    object 
 3   bill_length_mm     344 non-null    float64
 4   bill_depth_mm      344 non-null    float64
 5   flipper_length_mm  344 non-null    float64
 6   body_mass_g        344 non-null    float64
 7   sex                344 non-null    object 
 8   year               344 non-null    int64  
dtypes: float64(4), int64(2), object(3)
memory usage: 24.3+ KB


<h2>Changing the Data type</h2>

In [20]:
penguins['species'] = penguins['species'].astype('category')
penguins['island'] = penguins['island'].astype('category')
penguins['sex'] = penguins['sex'].astype('category')
penguins['year'] = penguins['year'].astype('category')

In [21]:
penguins.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype   
---  ------             --------------  -----   
 0   rowid              344 non-null    int64   
 1   species            344 non-null    category
 2   island             344 non-null    category
 3   bill_length_mm     344 non-null    float64 
 4   bill_depth_mm      344 non-null    float64 
 5   flipper_length_mm  344 non-null    float64 
 6   body_mass_g        344 non-null    float64 
 7   sex                344 non-null    category
 8   year               344 non-null    category
dtypes: category(4), float64(4), int64(1)
memory usage: 15.4 KB


<h2>Exploratory Data Analytics</h2>
<hr>

<h3>Question 1(Example)</h3>
<h4>which specie has the heaviest body mass ?</p>

In [22]:
penguins['body_mass_g'].mean()

np.float64(4201.754385964912)

In [23]:
gentoo_df = penguins[penguins['species'] == 'Gentoo']

In [24]:
gentoo_df['body_mass_g'].mean()

np.float64(5068.965761177136)

In [25]:
Chinstrap_df = penguins[penguins['species'] == 'Chinstrap']

In [26]:
Chinstrap_df['body_mass_g'].mean()

np.float64(3733.0882352941176)

In [27]:
adelie_df = penguins[penguins['species'] == 'Adelie']

In [28]:
adelie_df['body_mass_g'].mean()

np.float64(3703.95891043398)

<h4>Insight No.1</h4>
<p>Gentoo has the heaviest average body mass </p>

In [29]:
penguins['mass_kg'] = np.round(penguins['body_mass_g'] / 1000, 1)

In [30]:
penguins

rowid    species     island  bill_length_mm  bill_depth_mm  \
0        1     Adelie  Torgersen        39.10000       18.70000   
1        2     Adelie  Torgersen        39.50000       17.40000   
2        3     Adelie  Torgersen        40.30000       18.00000   
3        4     Adelie  Torgersen        43.92193       17.15117   
4        5     Adelie  Torgersen        36.70000       19.30000   
..     ...        ...        ...             ...            ...   
339    340  Chinstrap      Dream        55.80000       19.80000   
340    341  Chinstrap      Dream        43.50000       18.10000   
341    342  Chinstrap      Dream        49.60000       18.20000   
342    343  Chinstrap      Dream        50.80000       19.00000   
343    344  Chinstrap      Dream        50.20000       18.70000   

     flipper_length_mm  body_mass_g     sex  year  mass_kg  
0           181.000000  3750.000000    male  2007      3.8  
1           186.000000  3800.000000  female  2007      3.8  
2           195.000000  3250.000000  female  2007      3.2  
3           200.915205  4201.754386    male  2007      4.2  
4           193.000000  3450.000000  female  2007      3.4  
..                 ...          ...     ...   ...      ...  
339         207.000000  4000.000000    male  2009      4.0  
340         202.000000  3400.000000  female  2009      3.4  
341         193.000000  3775.000000    male  2009      3.8  
342         210.000000  4100.000000    male  2009      4.1  
343         198.000000  3775.000000  female  2009      3.8  

[344 rows x 10 columns]

In [31]:
penguins['mass_kg'].max()

np.float64(6.3)

In [32]:
penguins[penguins['mass_kg'] == 6.3]

rowid species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
169    170  Gentoo  Biscoe            49.2           15.2              221.0   

     body_mass_g   sex  year  mass_kg  
169       6300.0  male  2007      6.3

<h2>Activity(7 Questions & Insights)</h2>

<h3>Question 1</h3>
<h4>What is the maximum body mass of female penguins?</p>

In [33]:
penguins[penguins['sex'] == 'female']['body_mass_g'].max()

np.float64(5200.0)

<h4>Insight No.1</h4>
<p>The maximum body mass of female penguins is 5200.</p>

<h3>Question 2</h3>
<h4>What is the average bill depth for the Adelie species?</p>

In [34]:
penguins[penguins['species'] == 'Adelie']['bill_depth_mm'].mean()

np.float64(18.338494536780548)

<h4>Insight No.2</h4>
<p>The average bill depth of Adelie species is 18.3</p>

<h3>Question 3</h3>
<h4>What is the average body mass of male penguins from the 'Torgersen' island?</p>

In [35]:
penguins[(penguins['sex'] == 'male') & (penguins['island'] == 'Torgersen')]['body_mass_g'].mean()

np.float64(3990.2412280701756)

<h4>Insight No.3</h4>
<p>The average body mass of male penguins from the Torgersen island is 3990.</p>

<h3>Question 4</h3>
<h4>How many female penguins in year 2008?</p>

In [36]:
len(penguins[(penguins['sex'] == 'female') & (penguins['year'] == 2008)])

56

<h4>Insight No.4</h4>
<p>There are 56 Female penguins in year 2008.</p>

<h3>Question 5</h3>
<h4>How many penguins are in 3.5 below mass kg in Dream?</p>

In [37]:
len(penguins[(penguins['island'] == 'Dream') & (penguins['body_mass_g'] <= 3500)])

44

<h4>Insight No.5</h4>
<p>There are 44 penguins in Dream that have mass kg of 3.5 below.</p>

<h3>Question 6</h3>
<h4>What is the maximum body mass recorded for penguins from the Biscoe island?</p>

In [38]:
penguins[penguins['island'] == 'Biscoe']['body_mass_g'].max()

np.float64(6300.0)

<h4>Insight No.6</h4>
<p>The maximum body mass recorded for penguins from the Biscoe island is 6300.</p>

<h3>Question 7</h3>
<h4>What is the average flipper length for female penguins from the Dream island?</p>

In [41]:
penguins[(penguins['sex'] == 'female') & (penguins['island'] == 'Dream')]['flipper_length_mm'].mean()

np.float64(190.01639344262296)

<h4>Insight No.7</h4>
<p>The average flipper length for Female penguins from the Dream island is 190</p>

<h3>Question 8</h3>
<h4>Among male Chinstrap penguins, what is the average bill depth and how does it compare to females?</h4>

In [48]:
chinstrap = penguins[penguins['species'] == 'Chinstrap']
male_avg = chinstrap[chinstrap['sex'] == 'male']['bill_depth_mm'].mean()
female_avg = chinstrap[chinstrap['sex'] == 'female']['bill_depth_mm'].mean()

male_avg, female_avg

(np.float64(19.25294117647059), np.float64(17.588235294117652))

<h4>Insight No.8</h4>
<p>Male Chinstrap penguins have an average bill depth of 19.25 mm, while females have 17.58 mm a slight difference, possibly due to natural sexual dimorphism.</p>

<h3>Question 9</h3>
<h4>What is the average bill depth per sex for each island?</h4>

In [49]:
avg_bill_depth = penguins.groupby(['island', 'sex'])['bill_depth_mm'].mean()
avg_bill_depth

C:\Users\dreiz\AppData\Local\Temp\ipykernel_16600\1963636734.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  avg_bill_depth = penguins.groupby(['island', 'sex'])['bill_depth_mm'].mean()


island     sex   
Biscoe     female    15.191250
           male      16.510809
Dream      female    17.601639
           male      19.063492
Torgersen  female    17.550000
           male      19.137542
Name: bill_depth_mm, dtype: float64

<h4>Insight No.9</h4>
<p>This shows how bill depth differs by gender across islands. For instance, male penguins on Biscoe may have deeper bills than females which can indicate different feeding strategies or adaptations.</p>

<h3>Question 10</h3>
<h4>What is the total number of penguins per island and species combination?</h4>

In [50]:
penguin_counts = penguins.groupby(['island', 'species']).size().reset_index(name='count')
penguin_counts

C:\Users\dreiz\AppData\Local\Temp\ipykernel_16600\1717014980.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  penguin_counts = penguins.groupby(['island', 'species']).size().reset_index(name='count')


island    species  count
0     Biscoe     Adelie     44
1     Biscoe  Chinstrap      0
2     Biscoe     Gentoo    124
3      Dream     Adelie     56
4      Dream  Chinstrap     68
5      Dream     Gentoo      0
6  Torgersen     Adelie     52
7  Torgersen  Chinstrap      0
8  Torgersen     Gentoo      0

<h4>Insight No.10</h4>
<p>This helps you see where each species is most common. For example, most Gentoo penguins are found on Biscoe, while Chinstrap is mostly on Dream.</p>

<h3>Question 11</h3>
<h4>What is the difference in average body mass between male and female penguins for each species?</h4>

In [52]:
mass_by_sex_species = penguins.groupby(['species', 'sex'])['body_mass_g'].mean().unstack()
mass_diff = mass_by_sex_species['male'] - mass_by_sex_species['female']
mass_diff

C:\Users\dreiz\AppData\Local\Temp\ipykernel_16600\355644285.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  mass_by_sex_species = penguins.groupby(['species', 'sex'])['body_mass_g'].mean().unstack()


species
Adelie       644.794186
Chinstrap    411.764706
Gentoo       731.270051
dtype: float64

<h4>Insight No.11</h4>
<p>This shows the sex-based weight difference for each species. For example, Gentoo males may weigh about 500g more than females on average.</p>

<h3>Question 12</h3>
<h4>What is the mean flipper length for each year and species?</h4>

In [53]:
avg_flipper_by_year_species = penguins.groupby(['year', 'species'])['flipper_length_mm'].mean()
avg_flipper_by_year_species

C:\Users\dreiz\AppData\Local\Temp\ipykernel_16600\2019534550.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  avg_flipper_by_year_species = penguins.groupby(['year', 'species'])['flipper_length_mm'].mean()


year  species  
2007  Adelie       186.878304
      Chinstrap    192.423077
      Gentoo       215.117647
2008  Adelie       191.040000
      Chinstrap    197.722222
      Gentoo       217.565217
2009  Adelie       192.076923
      Chinstrap    198.083333
      Gentoo       218.020800
Name: flipper_length_mm, dtype: float64

<h4>Insight No.12</h4>
<p>You can track how flipper length changes over time by species. This might hint at evolutionary trends or measurement differences across research years.</p>

<h3>Question 13</h3>
<h4>What is the median flipper length for each species and sex?</h4>

In [55]:
median_flipper_by_species_sex = penguins.groupby(['species', 'sex'])['flipper_length_mm'].median()
median_flipper_by_species_sex

C:\Users\dreiz\AppData\Local\Temp\ipykernel_16600\1698788545.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  median_flipper_by_species_sex = penguins.groupby(['species', 'sex'])['flipper_length_mm'].median()


species    sex   
Adelie     female    188.0
           male      192.0
Chinstrap  female    192.0
           male      200.5
Gentoo     female    212.0
           male      221.0
Name: flipper_length_mm, dtype: float64

<h4>Insight No.13</h4>
<p>Median values are less affected by outliers. For example, male Gentoo penguins may have the highest median flipper length, while female Adelie penguins tend to have shorter flippers.</p>